스터디 그룹 정보 수정 API

kimshineday opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #122 
- 작업 요약: 어떤 작업을 했는지 간단하게 적어주세요.

## 📄 상세 내용
- [ ] 테스트 코드 작성 
    - 시리얼라이저 validate 및 update 테스트
    - API 테스트
- [ ] Serializer
    - 수정 시 검증을 위해 validate 조건 수정
    - update 작성
- [ ] views / urls
    - patch 함수 작성 및 라우터 추가

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

kimshineday committed:
✨ feat: Serializer update 함수 작성

- update 작성
- validate 검증 로직 보완
    - update시 검증 할 수 있게끔 변경
- profile_img 대한 정보가 validated_data에 없을 경우 생략하기 위해 if문으로 수정.

kimshineday committed:
✨ feat: 스터디 그룹 정보 수정 API 개발

- urls.py 라우터 추가
- UpdateStudyGroupView 작성

kimshineday committed:
✅ test: 스터디 그룹 정보 수정 기능 및 API 단위 테스트

- 시리얼라이저 validate 및 update 기능 테스트
- API 단위 테스트, patch 요청 테스트 진행

SimJunSik commented on apps/studies/serializers/study_group.py:
```diff
@@ -56,19 +56,23 @@ def validate(self, attrs: Dict[str, Any]) -> Dict[str, Any]:
+        if start_at is not None and end_at is not None:
+            # 시작일이 과거일 경우
+            if start_at < timezone.now():
+                raise serializers.ValidationError("스터디 시작일이 과거일 수 없습니다.")
```
여기 각 if 문 로직이 서로 바뀐 거 같네요

Meoyoug commented on apps/studies/serializers/study_group.py:
```diff
@@ -95,6 +100,31 @@ def create(self, validated_data: Dict[str, Any]) -> Any:
+            profile_img_url = s3_uploader.upload_file(file=img)
+            validated_data["profile_img_url"] = profile_img_url.get("url")
+
+        instance.name = validated_data.get("name", instance.name)
```
여기서 부터 부모 메서드를 활용해서 

```
with transaction.atomic():
    instance = super().update(validated_data)
    instance.lectures.set(lectures)
return instance
```

Meoyoug replied:
이렇게 줄여볼 수 있겠네요

Meoyoug replied:
트랜잭션을 활용해서 다른 필드의 업데이트가 실패할경우 lectures의 업데이트까지 롤백하도록 보장하구요

kimshineday committed:
💡 chore: validate if문 수정

- 반대로 작성된 if문 수정 완료

kimshineday committed:
♻️ refactor: transaction 사용, 중복 코드 제거

- transaction.atomic 사용, 중복된 코드 제거.

kimshineday committed:
💡 chore: 주석 추가 및 코드 포매터 적용 변경사항 커밋

- Serializer 주석 추가 및 수정
- 코드 포맷터 적용, 변경사항 저장

kimshineday committed:
✅ test: 인원 수 테스트 케이스 추가 및 코드 정리

- 테스트 케이스에 인원 수 테스트 추가
- error print문을 logger.debug로 변경
-

Meoyoug commented on apps/studies/tests/test_update_studygroup.py:
```diff
@@ -0,0 +1,162 @@
+        )  # 좀 더 좋은 방법이 있을거 같은데 모르겠음.
+        self.url = reverse("update_study_group", kwargs={"group_uuid": self.study_group.uuid})
+
+    def test_patch(self) -> None:
```
성공 케이스와 실패 케이스는 분리된 test 메서드로 작성합니다

각각의 케이스에 대해서도 분리된 test 메서드로 작성부탁드려요

kimshineday replied:
class로 나누는 것이 아닌 함수로만 구분을 지어도 괜찮을까요...?

Meoyoug replied:
네 같은 API에 대한 실패케이스 분리라면 함수분리로하셔도 충분합니다

Meoyoug commented on apps/studies/urls/study_group.py:
```diff
@@ -1,7 +1,9 @@
 
 urlpatterns = [
     path("", CreateStudyGroupView.as_view(), name="create_study_group"),
+    path("<uuid:group_uuid>/leader", UpdateStudyGroupView.as_view(), name="update_study_group"),
```
`<uuid:group_uuid>/leader`
이 엔드포인트의 의미는 뭔가요?

kimshineday replied:
그룹을 만든 리더만 수정할 수 있다고 해서 uuid 뒤에 leader 붙여서 구분하려고 했습니다..!

Meoyoug replied:
엔드포인트에는 굳이 명시하지 않아주셔도됩니다.

Meoyoug replied:
여기까지 수정되면 머지할게요~

kimshineday replied:
넵..! leader 부분 삭제 완료 했습니다..!

kimshineday committed:
♻️ refactor: 실행되지 않는 validate_max_headcount 삭제

- coverage 테스트에서 실행되지 않는 validate_max_headcount 삭제, StudyGroupCreateSerializer에서 IntegerField로 에러 메세지 지정.

kimshineday committed:
💡 chore: 테스트 코드에서 실패, 성공 케이스 코드 분리

- 기존 실패 성공 케이스를 한번에 테스트 진행에서 실패 성공 케이스를 함수로 분리.

kimshineday committed:
💡 chore: url 라우터 수정

- <uuid:group_uuid>/leader 에서 leader 삭제

Umdoong commented on apps/studies/serializers/study_group.py:
StudyGroupCreateSerializer의 field에 id를 uuid로 수정해주시고 extra kwargs로 readonly true 해주셔야 할 것 같습니다

kimshineday replied:
로직 안에서도 uuid를 사용해야하나요..?

Meoyoug replied:
로직안이라기보다 해당 시리얼라이저로 업데이트, 생성 뒤에 응답데이터 까지 같이만들어주는데 클라이언트가 내부에서 활용하기 위한용도인 id보다 uuid를 확인하고 활용할 수 있게 바꾸란것 같아요
fields 변수에 id -> uuid 로 변경하시고
extra_kwargs 활용해서 uuid를 read_only로만 해주시면 됩니다

Umdoong commented on apps/studies/serializers/study_group.py:
create와 update를 하나의 시리얼라이저에서 처리하고 계신데, 생성과 수정 시리얼라이저를 역할에 맞게 분리해주셔야 합니다.

kimshineday replied:
생성 항목과 수정 가능한 항목들이 같은데 시리얼라이저를 나눠서 처리하는게 좋은가요..?

Meoyoug replied:
굳이 분리할 필요는 없고 Serailizer 이름이 Create만 명시하고 있어서 Update도 추가해주시면 될것같아요

Umdoong commented on apps/studies/tests/test_update_studygroup.py:
```diff
@@ -0,0 +1,176 @@
+            self.assertFalse(serializer.is_valid())
+            logger.debug(serializer.errors)
+
+    def test_update_success(self) -> None:
```
unit test라서 인원수만 변경되는지, 이름만 변경되는지 따로 검증해야합니다.
그리고 save까지만 일어나고 있는데, save이후에도 결과값에 대한 검증도 추가해주셔야합니다.

kimshineday replied:
현재 가지고 있는 검증 로직 외에도 추가적으로 로직이 필요한건가요? 필요하다면 어떤 부분이 필요한건지 궁금합니다..!
저장 전에 검증을 하는데 저장 후에 어떤 검증이 필요한가요?

Meoyoug replied:
```
test_data: List[Dict[str, Any]] = [
            {"max_headcount": 5},
            {"profile_img": create_temp_image()},
            {"name": "이름 수정"},
        ]
```
이렇게 세가지 필드를 각각 업데이트하고 계신데 시리얼라이저를 통해 모델 업데이트 후에 각각의 업데이트 된 필드가 테스트 데이터와 일치하는지 확인하라는 것 같아요

Umdoong commented on apps/studies/views/study_group_update_view.py:
study group의 crud를 전부 분리하면 파일이 너무 많아질 것 같아요
간단한 crud는 하나의 view에서 관리해도 좋지 않을까요?

kimshineday replied:
crud를 나눠서 진행하고 있다보니 코드 작성 자체를 따로따로 작성하고 있습니다..!

Meoyoug replied:
이부분은 제가인지하고 있으니 추후에 완성하고 하나로 합쳐주시면 좋을듯해요

Umdoong commented on apps/studies/views/study_group_update_view.py:
```diff
@@ -0,0 +1,33 @@
+        study_group_data = get_object_or_404(StudyGroup, uuid=group_uuid)
+        serializer = StudyGroupCreateSerializer(instance=study_group_data, data=request.data, partial=True)
+        serializer.is_valid(raise_exception=True)
+        serializer.save(user=request.user)
```
update serializer를 따로 만들어서 분리해서 사용한다면,
update 내부에서 user를 사용하고 있지 않아서 굳이 user를 넘겨주지 않아도 될 것 같습니다.

Meoyoug replied:
여기도 시리얼라이저 분리안하셔도되니 user=request.user 만 지워주시면 됩니다!
Relation 필드인 유저를 업데이트하진 않으니꺄요

kimshineday committed:
💡 chore: StudyGroupCreateSerializer 필드 수정

- id를 uuid로 변경 후 read_only 설정

kimshineday committed:
💡 chore: Serializer 이름 변경

- StudyGroupCreateSerializer -> StudyGroupCreateUpdateSerializer 변경
- 시리얼라이저 이름 변경에 따른 다른 로직 파일에서의 반영

kimshineday committed:
💡 chore: patch시 save에서 request.user 삭제

- 수정에서 유저 정보가 변경 되지 않아 삭제.

kimshineday committed:
✨ feat: StudyGroupLeader Permission 작성

- 스터디 그룹 리더 권한 검증로직 추가
    - BasePermission 사용
    - Safe Methods로 들어가는 Get, Head, Options은 접근 가능
    - 특정 기능 접근 시 리더 권한이 없을 경우 403 에러 발생.

kimshineday committed:
✨ feat: 스터디 그룹 정보 수정 요청 시 그룹 리더 권한 유무 확인

- UpdateStudyGroupView에 permission_classes로 리더권한 유무 확인 추가
    - IsStudyGroupLeaderPermission

kimshineday committed:
✅ test: 리더 권한이 없는 경우 수정 실패 API 테스트 작성

- 새로운 유저를 생성, 권한 없는 유저가 수정 요청 시 403 에러 발생 하는지 확인.
    - 403 에러
    - 데이터가 수정이 되지 않아야함

kimshineday committed:
🎨 style: 코드 포매터 실행

- isort, black, mypy, coverage 실행

Meoyoug commented on apps/studies/permissions.py:
```diff
@@ -0,0 +1,38 @@
+        :param obj: StudyGroup 인스턴스
+        :return:
+        """
+        if request.method in permissions.SAFE_METHODS:  # SAFE_METHODS = GET, HEAD, OPTIONS
```
그룹 리더만 해당 오브젝트에 접근할 수 있게하는 퍼미션인 것 같은데 SAFE_METHODS인 경우에는 허락해주면 안되지 않나요..?

kimshineday replied:
일단 GET 요청 관련인거 같아 넣어두긴 했는데 한번 정보를 더 찾아보겠습니다..!

Meoyoug commented on apps/studies/permissions.py:
```diff
@@ -0,0 +1,38 @@
+            return False
+
+        # 리더 권한 확인
+        try:
```
여기는 
```
return study_group.members.exists(user=request.user, is_learder=True)
```

get 조회랑 exists 조회의 차이에대해 공부해보시면 좋을 것 같아요

kimshineday replied:
넵 수정 후 내용 관련해서 찾아보겠습니다..!

kimshineday committed:
♻️ refactor: IsStudyGroupLeaderPermission 리팩토링

- 리더 권한 확인을 위해 기존의 try문을 통한 확인 로직을 exists 사용으로 코드 수정

kimshineday committed:
💡 chore: IsStudyGroupLeaderPermission에서 SAFE_METHODS 삭제

- 온전히 스터디 그룹 리더에게만 권한을 부여하기 위해 허용 삭제.

Meoyoug reviewed (approved):
고생하셨습니다 머지할게용
